### PR TITLE
New scheduler int capacities

### PR DESCRIPTION
**Pull request by istoica**
Used fixed point instead of floating point (double) for scheduler resource operations. This should avoid corner cases where error accumulated using floating point operations may render a feasible allocation, as infeasible. 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 1]
```diff
@@ -629,7 +686,7 @@ int64_t ClusterResourceScheduler::NumNodes() { return nodes_.size(); }
 
 void ClusterResourceScheduler::UpdateResourceCapacity(const std::string &client_id_string,
                                                       const std::string &resource_name,
-                                                      int64_t resource_total) {
+                                                      double resource_total) {
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:689]: should make this `FixedPoint`

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -976,30 +1035,36 @@ void ClusterResourceScheduler::FreeTaskResourceInstances(
 
 std::vector<double> ClusterResourceScheduler::AddCPUResourceInstances(
     std::vector<double> &cpu_instances) {
+  std::vector<FixedPoint> cpu_instances_fp =
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:1038]: Can we have this function be
```
std::vector<FixedPoint> ClusterResourceScheduler::AddCPUResourceInstances(
    std::vector<FixedPoint> &cpu_instances)
```
?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 3]
```diff
@@ -976,30 +1035,36 @@ void ClusterResourceScheduler::FreeTaskResourceInstances(
...
   auto overflow = AddAvailableResourceInstances(
-      cpu_instances, &local_resources_.predefined_resources[CPU]);
+      cpu_instances_fp, &local_resources_.predefined_resources[CPU]);
   UpdateLocalAvailableResourcesFromResourceInstances();
 
-  return overflow;
+  return VectorFixedPointToVectorDouble(overflow);
 }
 
 std::vector<double> ClusterResourceScheduler::SubtractCPUResourceInstances(
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:1053]: Same as above

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 4]
```diff
@@ -30,23 +30,97 @@ enum PredefinedResources { CPU, MEM, GPU, TPU, PredefinedResources_MAX };
 // Specify resources that consists of unit-size instances.
 static std::unordered_set<int64_t> UnitInstanceResources{CPU, GPU, TPU};
 
-// Helper function to compare two vectors with double values.
-bool EqualVectors(const std::vector<double> &v1, const std::vector<double> &v2);
+/// Fixed point data type.
+class FixedPoint {
+#define RESOURCE_UNIT_SCALING 1000
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.h:35]: Should use a `constexpr` here instead

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 5]
```diff
@@ -39,6 +39,26 @@ std::string UnorderedMapToString(const std::unordered_map<std::string, double> &
   return buffer.str();
 }
 
+/// Convert a vector of doubles to a vector of resource units.
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:42]: Can we remove the below two functions and not use `std::vector<double>` for resources at all?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 6]
```diff
@@ -559,17 +603,24 @@ bool ClusterResourceScheduler::SubtractNodeAvailableResources(
     return false;
   }
 
+  FixedPoint zero(0.);
+
   for (size_t i = 0; i < PredefinedResources_MAX; i++) {
     resources.predefined_resources[i].available =
-        std::max(0., resources.predefined_resources[i].available -
-                         task_req.predefined_resources[i].demand);
+        resources.predefined_resources[i].available -
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:610]: I think `std::max(FixedPoint(0), resources.predefined_resources[i].available - task_req.predefined_resources[i].demand);` would be clearer here.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 7]
```diff
@@ -559,17 +603,24 @@ bool ClusterResourceScheduler::SubtractNodeAvailableResources(
...
+      resources.predefined_resources[i].available = zero;
+    }
   }
 
   for (const auto &task_req_custom_resource : task_req.custom_resources) {
     auto it = resources.custom_resources.find(task_req_custom_resource.id);
     if (it != resources.custom_resources.end()) {
-      it->second.available =
-          std::max(0., it->second.available - task_req_custom_resource.demand);
+      it->second.available = it->second.available - task_req_custom_resource.demand;
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:620]: Same as above.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 8]
```diff
@@ -592,16 +643,22 @@ bool ClusterResourceScheduler::AddNodeAvailableResources(int64_t node_id,
 
   for (size_t i = 0; i < PredefinedResources_MAX; i++) {
     resources.predefined_resources[i].available =
-        std::min(resources.predefined_resources[i].available +
-                     task_req.predefined_resources[i].demand,
-                 resources.predefined_resources[i].total);
+        resources.predefined_resources[i].available +
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:646]: Same as above.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 9]
```diff
@@ -592,16 +643,22 @@ bool ClusterResourceScheduler::AddNodeAvailableResources(int64_t node_id,
...
+          resources.predefined_resources[i].total;
+    }
   }
 
   for (const auto &task_req_custom_resource : task_req.custom_resources) {
     auto it = resources.custom_resources.find(task_req_custom_resource.id);
     if (it != resources.custom_resources.end()) {
-      it->second.available = std::min(
-          it->second.available + task_req_custom_resource.demand, it->second.total);
+      it->second.available = it->second.available + task_req_custom_resource.demand;
```
Comment (pcmoritz) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc:658]: Ditto.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 1 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc:689]: Actually the goal her was to not change the API in node_manager.cc

Then we need to do the mapping between double and FixedPoint in node_manager.cc.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 3 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc:1053]: Same as above.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 6 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc:610]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 7 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc:620]: done

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**